src/BaseControllerV2.ts

```diff
@@ -153,10 +155,26 @@ export class BaseController<S extends Record<string, any>> {
 }
 
 // This function acts as a type guard. Using a `typeof` conditional didn't seem to work.
-function isAnonymizingFunction<T>(x: boolean | Anonymizer<T>): x is Anonymizer<T> {
+function isAnonymizingFunction<T>(x: boolean | Anonymizer<T> | StateMetadata<T>): x is Anonymizer<T> {
   return typeof x === 'function';
 }
 
+function isStateMetadata<T>(x: StatePropertyMetadata<T> | StateMetadata<T>): x is StateMetadata<T> {
+  return Object.keys(x).sort() !== ['anonymous', 'persist'];
+}
+
+function isStatePropertyMetadata<T>(x: StatePropertyMetadata<T> | StateMetadata<T>): x is StatePropertyMetadata<T> {
+  return Object.keys(x).sort() === ['anonymous', 'persist'];
+}
+
+function isPrimitive(x: any): x is Primitive {
+  return ['boolean', 'string', 'number', 'null'].includes(typeof x);
+}
+
+function isNotArrayLike<T>(x: any): x is NotArrayLike<T> {
```

Comment:
All of these type guards were attempts at resolving the remaining type error, because apparently TypeScript needs a lot of help ruling out conditions in conditional types sometimes. These got rid of _some_ errors, but one still remains.